LLMs PR

india-kerle opened the pull request:
This PR closes the LangChain tutorial. I've started on another innovation mapping using LLMs tutorial but will add that in a separate PR so I can pivot to timeseries. 

To review the PR, please go through the notebook - some cells are incomplete purposefully so you can't run everything. Please make sure that the language makes sense and the tasks make sense! Follow along with the README.md to ensure you have added an OpenAI key as an environment variable.

Commit:
add basic structure for llm tutorial

Commit:
edit text

Commit:
add notes to self

Commit:
add armenian job adverts

Commit:
update data getters with armenian job ads

Commit:
add llm utils

Commit:
add llm utils

Commit:
add llm utils

Commit:
add llm utils

Commit:
update langchain tutorial

Commit:
fix typos

Commit:
fix typo

Commit:
fix typo

Commit:
just opt for basics of langchain

Commit:
deal with DG merge conflict

india-kerle committed:
Merge branch 'dev' into 1_llms

Commit:
rename utils

Commit:
edit readmes for consistency

georgerichardson reviewed (commented):
- Few-shot prompting is mentioned before it's introduced in the `## Prompts` cell
- I think there's room for minor confusion over what to do or participants deleting the `[]` in the 'Can you create a similar prompt using the `ChatPromptTemplate`' task.
- I bet you get a question about using `**` under the first example of few shot prompting.
- Whan you discuss selectors, maybe it's worth spelling out what they are briefly and then asking why people might use them, rather than diverting people to the docs.
- Lots of unclosed parentheses in the second few shot prompt code cell.
- Why might you use stuff, map_reduce and refine for index related chains?
- I don't get any of the Kardashian references. This is probably my problem 👴 
- There's quite a lot to cover. Depending on how you're doing for time, you could outline the last few advanced use cases and then let people explore them at will in parallel.

india-kerle commented:
Actions:

- [x] Deleted mention of few-shot prompting for clarification 
- [x] Added more comments in the code to tell participants to delete [] in the 'Can you create a similar prompt using the ChatPromptTemplate' task.
- [x] Added a comment in the code describing the role of the ** operator in unpacking dictionaries 
- [x] Define selectors in markdown before asking people why they might use them 
- [x] re: unclosed parentheses -> I'm trying to give people pseudo-code so i've just spelt it out more! 
- [x] Added more context around why you would use stuff, map_reduce + refine 
- [x] devos re: kardashians 
- [x] Agreed re: alot to cover. I've added that as a note to mention in the slide deck!

Commit:
make george's changes
